src/com/goide/editor/GoStatementMover.java

```diff
@@ -0,0 +1,143 @@
+
+
+public class GoStatementMover extends LineMover {
+  @Override
+  public boolean checkAvailable(@NotNull Editor editor, @NotNull PsiFile file, @NotNull MoveInfo info, boolean down) {
+    if (!(file instanceof GoFile && super.checkAvailable(editor, file, info, down))) return false;
+
+    Pair<PsiElement, PsiElement> primeElementRangePair = getElementRange(editor, file, getLineRangeFromSelection(editor));
+    if (primeElementRangePair == null) return false;
+    Couple<PsiElement> primeElementRange = new Couple<PsiElement>(primeElementRangePair.first, primeElementRangePair.second);
+    PsiElement commonParent =
+      primeElementRange.first.isEquivalentTo(primeElementRange.second) ? primeElementRange.first.getParent() :
+      PsiTreeUtil.findCommonParent(primeElementRange.first, primeElementRange.second);
+    if (commonParent == null) return false;
+
+    Couple<PsiElement> elementRange = getLogicElementRange(primeElementRange, commonParent);
+    if (elementRange == null) return false;
+    if (commonParent == elementRange.first) commonParent = commonParent.getParent();
+    info.toMove = new LineRange(elementRange.first, elementRange.second);
+    if (elementRange.first instanceof GoTopLevelDeclaration) {
+      GoTopLevelDeclaration toMove2 =
+        (GoTopLevelDeclaration)getElementOfTypeAfter(commonParent, elementRange, GoTopLevelDeclaration.class, down);
+      info.toMove2 = toMove2 == null ? null : new LineRange(toMove2);
+      return true;
+    }
+    if (commonParent instanceof GoImportList) {
+      GoImportDeclaration toMove2 =
+        (GoImportDeclaration)getElementOfTypeAfter(commonParent, elementRange, GoImportDeclaration.class, down);
+      info.toMove2 = toMove2 == null ? null : new LineRange(toMove2);
+      return true;
+    }
+    return setUpInfo(info, elementRange, commonParent, down);
+  }
+
+  @Nullable
+  private static Couple<PsiElement> getLogicElementRange(@NotNull Couple<PsiElement> elementRange,
+                                                         @NotNull PsiElement parent) {
+    if (elementRange.first == null || elementRange.second == null) return null;
+    int start = elementRange.first.getTextOffset();
+    int end = elementRange.second.getTextRange().getEndOffset();
+
+    TextRange range = parent.getTextRange();
+    PsiElement[] children = parent.getChildren();
+    if (range.getStartOffset() == start && (children.length == 0 || children[0].getTextOffset() > start) ||
+        range.getEndOffset() == end && (children.length == 0 || children[children.length - 1].getTextRange().getEndOffset() < end)) {
+      return new Couple<PsiElement>(parent, parent);
+    }
+
+    PsiElement startElement = elementRange.first;
+    PsiElement endElement = elementRange.second;
+    for (PsiElement element : children) {
+      range = element.getTextRange();
+      if (range.getStartOffset() <= start && range.getEndOffset() <= end && range.getEndOffset() > start) {
```

Comment:
Add explanation in comment
    // get range which contains TextRange(start, end) of top level elements
    // common parent of elements is straight parent for each element